parsers/ELEXON.py

```diff
@@ -84,17 +88,121 @@
+        )
+    return r.json()
+
+
+def parse_datetime(settlementDate: str, settlementPeriod: int) -> datetime:
+    parsed_datetime = datetime.strptime(settlementDate, "%Y-%m-%d")
+    parsed_datetime += timedelta(hours=(settlementPeriod - 1) / 2)
+    return parsed_datetime.replace(tzinfo=timezone.utc)
+
+
+def query_production(
+    session: Session, target_datetime: datetime, logger: Logger
+) -> ProductionBreakdownList:
+    """Fetches production data from the B1620 endpoint from the ELEXON API."""
+    production_params = {
+        "publishDateTimeFrom": (target_datetime - timedelta(days=2)).strftime(
+            "%Y-%m-%d 00:00"
+        ),
+        "publishDateTimeTo": target_datetime.strftime("%Y-%m-%d %H:%M"),
+    }
+    production_data = query_elexon(
+        ELEXON_URLS["production"], session, production_params
+    )
+
+    parsed_events = parse_production(production_data, logger, "B1620")
+    return parsed_events
+
+
+def get_event_value(event: dict[str, Any], key: str) -> float | None:
+    value = event.get(key)
+    if value is not None:
+        return float(value)
+
+
+def parse_production(
+    production_data: list[dict[str, Any]], logger: Logger, dataset: str
+) -> ProductionBreakdownList:
+    """Parses production events from the ELEXON API. This function is used for the B1620 data or the FUELHH data."""
+    dataset_info = {
+        "B1620": {
+            "mode_mapping": RESOURCE_TYPE_TO_FUEL,
+            "mode_key": "psrType",
+            "quantity_key": "quantity",
+        },
+        "FUELHH": {
+            "mode_mapping": FUEL_INST_MAPPING,
+            "mode_key": "fuelType",
+            "quantity_key": "generation",
+        },
+    }
+
+    mode_mapping = dataset_info[dataset]["mode_mapping"]
+    mode_key = dataset_info[dataset]["mode_key"]
+    quantity_key = dataset_info[dataset]["quantity_key"]
+
+    all_production_breakdowns: list[ProductionBreakdownList] = []
+
+    for event in production_data:
+        production_breakdown = ProductionBreakdownList(logger=logger)
+        event_datetime = parse_datetime(
+            event.get("settlementDate"), event.get("settlementPeriod")
+        )
+        production_mix = ProductionMix()
+        storage_mix = StorageMix()
+
+        production_mode = mode_mapping[event.get(mode_key)]
+
+        if production_mode == "exchange":
+            continue
+
+        if production_mode == "hydro storage":
+            storage_value = get_event_value(event, quantity_key)
+            if storage_value:
+                storage_mix.add_value("hydro", -1 * storage_value)
+                production_breakdown.append(
+                    zoneKey=ZoneKey("GB"),
+                    storage=storage_mix,
+                    source=ELEXON_SOURCE,
+                    datetime=event_datetime,
+                )
+        else:
+            production_value = get_event_value(event, quantity_key)
+            if production_value:
```

Comment:
This if clause is what is messing with the solar values since 0 is false but it's also filtering out all the 0 mess from the B1620 dataset.

Reply:
I've added a validation function further down that checks if the production breakdown is complete

Reply:
Does that mean this can be removed?

It's still causing the following issue with no solar at night:
```python
 {'correctedModes': [],
  'datetime': datetime.datetime(2024, 3, 27, 5, 30, tzinfo=datetime.timezone.utc),
  'production': {'biomass': 1816.0,
                 'coal': 270.0,
                 'gas': 2938.0,
                 'hydro': 345.0,
                 'nuclear': 4564.0,
                 'unknown': 418.0,
                 'wind': 14942.188},
  'source': 'elexon.co.uk, nationalgrideso.com',
  'sourceType': <EventSourceType.measured: 'measured'>,
  'storage': {'hydro': -135.0},
  'zoneKey': 'GB'},
 {'correctedModes': [],
  'datetime': datetime.datetime(2024, 3, 27, 6, 0, tzinfo=datetime.timezone.utc),
  'production': {'biomass': 2086.0,
                 'coal': 329.0,
                 'gas': 3589.0,
                 'hydro': 380.0,
                 'nuclear': 4582.0,
                 'solar': 78.0,
                 'unknown': 425.0,
                 'wind': 15143.47},
  'source': 'elexon.co.uk, nationalgrideso.com',
  'sourceType': <EventSourceType.measured: 'measured'>,
  'storage': {'hydro': 65.0},
  'zoneKey': 'GB'},
  ```